### PR TITLE
add [[CrossCol]] Operation

### PR DESCRIPTION
**Pull request by tosky001**
## What changes were proposed in this pull request?

add CrossCol Operation to describe the categorical feature interactions.

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by tosky001** [comment]
@yiheng this PR is for categorical feature crossing. looking forward your reviews and thanks.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    require(tensorNum>=2, "the input table must contain more than one tensor")
+    val batchSize = input[Tensor[String]](1).size(dim = 1)
+
+    val indices0 = new ArrayBuffer[Int]()
+    val indices1 = new ArrayBuffer[Int]()
+    val values = new ArrayBuffer[Int]()
+    val shape = Array(batchSize, hashBucketSize)
+
+    (1 to tensorNum).foreach(i =>
+      input[Tensor[String]](i).squeeze()
```
[the remark this answers is not in the document]
Comment (tosky001) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala]: the **squeeze** is aimed to handle the 1D or 2D(batch_size*1) input tensor. and the side effect has been modified by using the `Tensor.view()`.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+ *  "1", "2", "3,4"
+ *
+ *  the output tensor should be an 2D 3 x hashBucketSize SparseTensor:
+ *  [0, 0]: Hash32("1", Hash32("D")) % hashBucketSize
+ *  [0, 1]: Hash32("1", Hash32("A")) % hashBucketSize
+ *  [1, 0]: Hash32("2", Hash32("B")) % hashBucketSize
+ *  [2, 0]: Hash32("3", Hash32("C")) % hashBucketSize
+ *  [2, 1]: Hash32("4", Hash32("C")) % hashBucketSize
+ *  [2, 2]: Hash32("3", Hash32("A")) % hashBucketSize
+ *  [2, 3]: Hash32("4", Hash32("A")) % hashBucketSize
```
[the remark this answers is not in the document]
Comment (tosky001) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala]: sorry, it's my mistake and the `shape of output` has been modified.

**Comment by tosky001** [comment]
@yiheng the ops has been modified and run successfully.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * As for the SparseTensor, it should be transformed to
+ * Tensor[String] before feeding into the Operation
+ *
+ *
+ * For example, if the two input tensors with size=3 are:
+ *  "A,D", "B", "A,C"
+ *  "1", "2", "3,4"
+ *
+ *  the output tensor should be an 2D 3 x hashBucketSize SparseTensor:
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala]: Do we need to update the ```2D 3 x hashBucketSize``` to ```2D 3 x maxLength```?

**Comment by yiheng** [comment]
running jenkins

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala (hunk 3 above)
Comment (tosky001) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala]: @yiheng my mistake and the update has well done.

**Comment by yiheng** [comment]
style check failed at:
```
error file=/opt/work/jenkins/workspace/BigDL-PRVN-StyleCheck/spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/CrossCol.scala message=File must end with newline character
```

**Comment by tosky001** [comment]
@yiheng done.

**Comment by yiheng** [comment]
jenkins pass. Merged. Thanks for the contribution!

**Pull request merged**